### ISSUE TITLE
Feature Request: make `vue/no-undef-components` works in Vue class component

### ISSUE THREAD

**Issue by ylc395** [vuejs/eslint-plugin-vue#2546]
**What rule do you want to change?**
vue/no-undef-components

**Does this change cause the rule to produce more or fewer warnings?**
fewer warning

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behavior

**Please provide some example code that this change will affect:**

```vue
<template>
<div>
  <Foo /> <!-- This will trigger a vue/no-undef-components warning, which is incorrect-->
</div>
</template>
<script>
@Component({
  components: {
    Foo
  }
})
export default class Bar extends Vue {}
</script>
```

**What does the rule currently do for this code?**
trigger a vue/no-undef-components warning, which is incorrect

**What will the rule do after it's changed?**
do nothing

**Comment by ylc395** [comment]
If this feature request is acceptable, I can submit a PR.

**Comment by FloEdelmann** [comment]
If this is possible without much code, then yes, please submit a PR! Make sure to also include test cases (valid and invalid ones) to verify the implementation.

**Comment by ylc395** [comment]
I almost finish the PR, but I notice that just use comment like this:

```vue
<template>
  <Bar />
</template>
<script>
// @vue/component
@Component({
  components: { Bar } 
})
export default class Foo {}
</script>
```
will solve this problem.

So I think we do not need support Vue class component on purpose in this case

**Comment by FloEdelmann** [comment]
Feel free to open a PR to improve the docs!

**Comment by ylc395** [comment]
OK, please check #2561
